File: LeetCode/1122-relative-sort-array/1122-relative-sort-array.java
class Solution {
    public int[] relativeSortArray(int[] arr1, int[] arr2) {
        Map<Integer, Integer> mp = new HashMap<>();
        for (int i = 0; i < arr2.length; i++) {
            mp.put(arr2[i], i);
        }
        int largeval = 1000000000;
        for (int i = 0; i < arr1.length; i++) {
            if (!mp.containsKey(arr1[i])) {
                mp.put(arr1[i], largeval);
            }
        }
        Comparator<Integer> cmp = (num1, num2) -> {
            int i1 = mp.get(num1);
            int i2 = mp.get(num2);
            if (i1 == i2) {
                return Integer.compare(num1, num2);
            }
            return Integer.compare(i1, i2);
        };
        List<Integer> arr = new ArrayList<>();
        for (int num : arr1) {
            arr.add(num);
        }
        Collections.sort(arr, cmp);
        for (int i = 0; i < arr1.length; i++) {
            arr1[i] = arr.get(i);
        }
        return arr1;
    }
}



// class Solution {
//     public int[] relativeSortArray(int[] arr1, int[] arr2) {
//         TreeMap<Integer, Integer> mp = new TreeMap<>();
//         for (int x : arr1) {
//             mp.put(x, mp.getOrDefault(x, 0) + 1);
//         }
        
//         List<Integer> ans = new ArrayList<>();
//         for (int x : arr2) {
//             while (mp.get(x) > 0) {
//                 ans.add(x);
//                 mp.put(x, mp.get(x) - 1);
//             }
//             mp.remove(x);
//         }
        
//         for (Map.Entry<Integer, Integer> x : mp.entrySet()) {
//             int key = x.getKey();
//             int value = x.getValue();
//             while (value > 0) {
//                 ans.add(key);
//                 value--;
//             }
//         }
        
//         int[] result = new int[ans.size()];
//         for (int i = 0; i < ans.size(); i++) {
//             result[i] = ans.get(i);
//         }
        
//         return result;
//     }
// }

File: LeetCode/1122-relative-sort-array/1122-relative-sort-array.py
class Solution:
    def relativeSortArray(self, arr1: List[int], arr2: List[int]) -> List[int]:
        mp = {val: idx for idx, val in enumerate(arr2)}
        largeval = 1000000000
        for num in arr1:
            if num not in mp:
                mp[num] = largeval
        def cmp(num1, num2):
            i1 = mp[num1]
            i2 = mp[num2]
            if i1 == i2:
                return num1 - num2
            return i1 - i2
        arr1.sort(key = lambda x : (mp[x], x))
        return arr1


# class Solution:
#     def relativeSortArray(self, arr1: List[int], arr2: List[int]) -> List[int]:
#         mp = {}
#         for x in arr1:
#             if x in mp:
#                 mp[x] += 1
#             else:
#                 mp[x] = 1
#         ans = []
#         for x in arr2:
#             while mp[x] > 0:
#                 ans.append(x)
#                 mp[x] -= 1
#             del mp[x]
#         for k in sorted(mp.keys()):
#             while mp[k] > 0:
#                 ans.append(k)
#                 mp[k] -= 1
#         return ans